### ISSUE TITLE
14_recurrent_neural_networks.ipynb Machine Translation

### ISSUE THREAD

**Issue by jamesdhope** [ageron/handson-ml#82]
Dear Aurélien,

Again, thanks for your brilliant book.

Please would it be possible to show me how to run the RNN encoder-decoder In[112].

I understand W to be the weights for the decoder so that Cross Entropy only calculates losses for populated sequence values. In other words, the weights are 0 if the target Y is not populated (i.e.  EOS) and 1.0 if the target Y is populated (i..e it does contain a sequence value). 

I would be sincerely grateful if you could please elaborate on this a little further, and how you would go about implementing this model in at the end of your notebook, using batch gradient descent.

James

**Comment by ageron** [comment]
Hi @jamesdhope ,
Thanks for your kind words, and sorry for the late response. I have little time due to consulting, but I'll try to answer by the end of the week.

**Comment by jamesdhope** [comment]
Thanks. I look forward to hearing from you.

Best wishes
James

**Comment by ageron** [comment]
The code in the cells [112] and [113] is roughly inspired from the [Sequence-to-Sequence tutorial](https://www.tensorflow.org/tutorials/seq2seq), it was just there to give the general idea.

In short, as always the training loop should repeatedly run the `training_op` and feed it a training batch using the placeholders `X`, `Y` and `W`.  Each row in `X` must be an English sentence ("I drink milk"), tokenized (["I", "drink", "milk"]), reversed (["milk", "drink", "I"]), padded using a "<pad>" token (["<pad>", "<pad>", "milk", "drink", "I"]) and converted to integer word IDs ([0, 0, 132, 57, 240]). Each row in `Y` must represent the French translation, preprocessed similarly (except, not reversed). And finally `W` must have the same shape as `Y`, and it must be equal to 0.0 where `Y` is <pad>, or 1.0 otherwise.

But the devil is in the details, so I would recommend you go through the tutorial if you haven't already done so.

Hope this helps.
